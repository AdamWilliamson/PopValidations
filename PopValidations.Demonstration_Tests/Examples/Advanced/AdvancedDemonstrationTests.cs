using ApprovalTests;
using FluentAssertions;
using FluentAssertions.Execution;
using Newtonsoft.Json.Linq;
using PopValidations.Swashbuckle_Tests.Helpers;
using PopValidations_Tests.TestHelpers;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using Xunit;

namespace PopValidations.Demonstration_Tests.Examples.Advanced;

public class AdvancedDemonstrationTests
{
    [Fact]
    public async Task Validation()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(
            new AdvancedDemonstration.AlbumSubmissionValidator(
                new AdvancedDemonstration.AlbumVerificationService()
            )
        );
        var albumSubmission = new AdvancedDemonstration.AlbumSubmission(
            new() {
                new AdvancedDemonstration.Album(
                    "Down with the Sickness",
                     AdvancedDemonstration.AlbumType.SingleArtist,
                    new (){ new AdvancedDemonstration.Artist("") },
                    "https://en.wikipedia.org/wiki/File:TheSickness.jpg",
                    new System.DateTime(2000, 03, 07),
                    new()
                    {
                        new AdvancedDemonstration.Song(
                            new(){ new AdvancedDemonstration.Artist("Disturbed") },
                            4,
                            "Down With The Sickness",
                            4.38,
                            "Nu Metal"
                        ),
                        new AdvancedDemonstration.Song(
                            new() { new AdvancedDemonstration.Artist("Muse") },
                            1,
                            "Supremecy",
                            4.51,
                            "Progressive Rock"
                        ),
                        null
                    },
                    new (){ "Rock", "Nu Metal" }
                ),
                new AdvancedDemonstration.Album(
                    "",
                     AdvancedDemonstration.AlbumType.Collaboration,
                    new() { new AdvancedDemonstration.Artist("Disturbed") },
                    null,
                    null,
                    new()
                    {
                        new AdvancedDemonstration.Song(
                            null,
                            null,
                            "",
                            -1,
                            "Pop"
                            ),
                        null
                    },
                    new() { "Rock", "Nu Metal" }
                ),
                new AdvancedDemonstration.Album(
                    "",
                     AdvancedDemonstration.AlbumType.Compilation,
                    new() { new AdvancedDemonstration.Artist("Disturbed") },
                    null,
                    null,
                    new()
                    {
                        new AdvancedDemonstration.Song(
                            new() { new AdvancedDemonstration.Artist("Disturbed") },
                            null,
                            "",
                            -1,
                            "Rock"
                        ),
                        new AdvancedDemonstration.Song(
                            new() { new AdvancedDemonstration.Artist("Disturbed") },
                            null,
                            "",
                            -1,
                            "Country"
                        ),
                    },
                    new() { "Pop" }
                ),
                new AdvancedDemonstration.Album(
                    "",
                     AdvancedDemonstration.AlbumType.Single,
                    new() { new AdvancedDemonstration.Artist("Disturbed") },
                    null,
                    null, 
                    Enumerable.Repeat(0, 8).Select<int, AdvancedDemonstration.Song?>(i => 
                        new AdvancedDemonstration.Song(
                            null,
                            null,
                            "",
                            -1,
                            "Rock"
                        )
                    ).ToList(),
                    new() { "Pop" }
                ),
                new AdvancedDemonstration.Album(
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null
                ),
                null
            }
        );

        // Act
        var results = await runner.Validate(albumSubmission);
        var json = JsonConverter.ToJson(results);

        // Assert
        Approvals.VerifyJson(json);

        void AssertType(List<string> parentNames, Type t, List<string> ignore)
        {
            if (t.IsGenericType && typeof(List<>) == t.GetGenericTypeDefinition()) 
            { 
                t = t.GetGenericArguments()[0];
            }

            foreach (var prop in t.GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly))
            {
                if (
                    prop.PropertyType.DeclaringType?.Name == nameof(AdvancedDemonstration)
                    || (
                        prop.PropertyType.IsGenericType
                        && prop.PropertyType.GetGenericArguments()[0].DeclaringType?.Name == nameof(AdvancedDemonstration)
                    )
                )
                {
                    var newListOfParents = new List<string>(parentNames);
                    newListOfParents.Add(prop.Name);
                    if (!prop.PropertyType.IsEnum && (!Nullable.GetUnderlyingType(prop.PropertyType)?.IsEnum ?? true))
                    {
                        AssertType(newListOfParents, prop.PropertyType, ignore);
                    }
                    results.GetErrorsForField(newListOfParents).Should().HaveCountGreaterThan(0, string.Join('.', newListOfParents));
                }
            }
        }

        using (new AssertionScope()) 
        {
            AssertType(new() { }, typeof(AdvancedDemonstration.AlbumSubmission), new() { "Albums" });
        }
    }

    [Fact]
    public void Description()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(
            new AdvancedDemonstration.AlbumSubmissionValidator(new AdvancedDemonstration.AlbumVerificationService())
        );

        // Act
        var results = runner.Describe();
        var json = JsonConverter.ToJson(results);

        // Assert
        Approvals.VerifyJson(json);
    }

    [Fact]
    public async Task OpenApi()
    {
        // Arrange
        var config = new WebApiConfig();
        var setup = new TestSetup<
               AdvancedDemonstration.TestController,
               AdvancedDemonstration.AlbumSubmissionValidator,
               AdvancedDemonstration.AlbumSubmission
           >(new AdvancedDemonstration.AlbumSubmissionValidator(new AdvancedDemonstration.AlbumVerificationService()));
        setup.Register(typeof(AdvancedDemonstration.AlbumVerificationService), new AdvancedDemonstration.AlbumVerificationService());
        // Act
        var helper = await setup.GetHelper(config);
        JObject json = JObject.Parse(helper.Content!);
        var match = json["components"]!["schemas"]!;//.FirstOrDefault();

        // Assert
        Approvals.VerifyJson(match?.ToString(Newtonsoft.Json.Formatting.None));
    }
}

