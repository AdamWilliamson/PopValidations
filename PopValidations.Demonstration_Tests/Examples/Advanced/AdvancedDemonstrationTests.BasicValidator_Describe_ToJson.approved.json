{
    "results": [
        {
            "property": "Artist.Name",
            "outcomes": [],
            "validationGroups": [
                {
                    "name": "Artist is not null",
                    "description": "Artist is not null",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "NotNullValidation",
                            "message": "Must not be null.",
                            "values": []
                        },
                        {
                            "validator": "IsEqualToValidation",
                            "message": "Must equal to ''artist' with Something appended'",
                            "values": [
                                {
                                    "key": "value",
                                    "value": "'artist' with Something appended"
                                }
                            ]
                        }
                    ]
                }
            ]
        },
        {
            "property": "Created",
            "outcomes": [
                {
                    "validator": "IsCustomScopedValidation",
                    "message": "Name is too short",
                    "values": []
                }
            ],
            "validationGroups": []
        },
        {
            "property": "Songs[n]",
            "outcomes": [
                {
                    "validator": "NotNullValidation",
                    "message": "Must not be null.",
                    "values": []
                }
            ],
            "validationGroups": [
                {
                    "name": "Album constains songs we don't own the copyright to",
                    "description": "Album constains songs we don't own the copyright to",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "NotNullValidation",
                            "message": "Must not be null.",
                            "values": []
                        },
                        {
                            "validator": "IsCustomScopedValidation",
                            "message": "Checks to ensure the song's rights are owned by us.",
                            "values": []
                        }
                    ]
                }
            ]
        },
        {
            "property": "Songs[n].TrackNumber",
            "outcomes": [
                {
                    "validator": "NotNullValidation",
                    "message": "Must not be null.",
                    "values": []
                }
            ],
            "validationGroups": []
        },
        {
            "property": "Songs[n].TrackName",
            "outcomes": [
                {
                    "validator": "IsEqualToValidation",
                    "message": "Must equal to 'Definitely Not The Correct Song Name.'",
                    "values": [
                        {
                            "key": "value",
                            "value": "Definitely Not The Correct Song Name."
                        }
                    ]
                }
            ],
            "validationGroups": []
        },
        {
            "property": "Songs[n].Duration",
            "outcomes": [
                {
                    "validator": "IsEqualToValidation",
                    "message": "Songs must force you to travel slowly backwards in time to listen to.",
                    "values": [
                        {
                            "key": "value",
                            "value": "-1"
                        }
                    ]
                }
            ],
            "validationGroups": [
                {
                    "name": "Song Duration is greater than 5 minutes",
                    "description": "Song Duration is greater than 5 minutes",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "IsEqualToValidation",
                            "message": "Must equal to '10.1'",
                            "values": [
                                {
                                    "key": "value",
                                    "value": "10.1"
                                }
                            ]
                        }
                    ]
                }
            ]
        },
        {
            "property": "Songs[n].Artist",
            "outcomes": [],
            "validationGroups": [
                {
                    "name": "Album is Single Artist",
                    "description": "Album is Single Artist",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "IsNullValidation",
                            "message": "Must be null.",
                            "values": []
                        }
                    ]
                },
                {
                    "name": "Album is SingleArtist",
                    "description": "Album is SingleArtist",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "IsNullValidation",
                            "message": "Must be null.",
                            "values": []
                        }
                    ]
                }
            ]
        },
        {
            "property": "Artist",
            "outcomes": [],
            "validationGroups": [
                {
                    "name": "Album is Single Artist",
                    "description": "Album is Single Artist",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "NotNullValidation",
                            "message": "Must not be null.",
                            "values": []
                        }
                    ]
                }
            ]
        },
        {
            "property": "Songs[n].Artist.Name",
            "outcomes": [],
            "validationGroups": [
                {
                    "name": "Album is Collaboration",
                    "description": "Album is Collaboration",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "NotNullValidation",
                            "message": "Must not be null.",
                            "values": []
                        }
                    ]
                },
                {
                    "name": "Album is Collaborative",
                    "description": "Album is Collaborative",
                    "children": [],
                    "outcomes": [
                        {
                            "validator": "NotNullValidation",
                            "message": "Must not be null.",
                            "values": []
                        }
                    ]
                }
            ]
        }
    ]
}